fendermint/vm/interpreter/src/fvm/gas_estimation.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::Result;
+
+use crate::fvm::{
+    observe::{MsgExec, MsgExecPurpose},
+    state::FvmQueryState,
+};
+use fendermint_vm_message::query::GasEstimate;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::{self, RawBytes};
+use fvm_shared::{
+    bigint::BigInt, econ::TokenAmount, error::ExitCode, message::Message, BLOCK_GAS_LIMIT,
+};
+use ipc_observability::emit;
+use num_traits::Zero;
+use std::time::Instant;
+
+/// Estimates the gas for a given message.
+pub async fn estimate_gassed_msg<DB: Blockstore + Clone + 'static + Send + Sync>(
+    state: FvmQueryState<DB>,
```

Comment:
a ref of state will work?

Reply:
Unfortunately not, because it uses RefCell<std::option::Option<FvmExecState<ReadOnlyBlockstore<DB>>>> which won't work in `async`.